system/lib/wasmfs/file_table.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include <assert.h>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include <wasi/api.h>
+
+class OpenFileDescriptor {
+  std::shared_ptr<File> file;
+  __wasi_filedelta_t offset;
+  std::mutex mutex;
+
+public:
+  OpenFileDescriptor(uint32_t offset, std::shared_ptr<File> file);
+
+  std::shared_ptr<File>& getFile();
+};
+
+class FileTable {
+  static std::vector<std::shared_ptr<OpenFileDescriptor>> entries;
+  std::mutex mutex;
+
+  FileTable() {
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdinFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdoutFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StderrFile>()));
+  }
+
+public:
+  // Handle represents an RAII wrapper object. Access to the FileTable must go through a Handle.
+  // A Handle holds FileTable's lock for the duration of its lifetime.
```

Comment:
Two downsides of the Handle approach come to mind:

* We need to write all FileTable operations on the Handle, or pass through to them, which is less concise and obvious.
* If we want to refine how we do locking later, the structure of "take a lock for an entire access to the table" would be deeply embedded in the code, and hard to remove.

Instead, how about having FileTable have methods directly, like `openFile()`, and those methods would take an RAII lock?

Reply:
That was the original design, but it doesn't work because there are cases (such as in dup2) where the different file table methods have to be combined into single atomic operations, so the lock has to be held across multiple methods. Without the `Handle`, the only way to do that would be to make the callers of file table methods responsible for locking, but that's extremely error prone since they would have to separately remember to take a lock for every access.

With the operator overloading and implicit conversions, the handle approach is also much more concise than any external locking scheme.

The handle approach is also compatible with more fine-grained locking, since the handle to the file table can be dropped (and its lock released) once any top-level accesses to the file table are complete. In other words, the handle lock only needs to protect access to the table itself, not necessarily access to the `OpenFileDescriptor`s or `File`s.

Reply:
I see, thanks, makes sense.

Sgtm. Let's document this on a comment on Handle, unless it's already there and I missed it?

Reply:
Sounds good, will add it to a comment